lib/commands/skill/add-locales/helper.js

```diff
@@ -0,0 +1,148 @@
+ * and commit the copy paste from the source to the target location.
+ *
+ * @param {Array} selectedLocales
+ * @param {String} profile
+ * @param {Boolean} doDebug
+ * @param {Function} callback
+ */
+function addLocales(selectedLocales, profile, doDebug, callback) {
+    _getNewLocaleModelUri(selectedLocales, profile, doDebug, (resolveErr, iModelSourceByLocales) => {
+        if (resolveErr) {
+            return callback(resolveErr);
+        }
+        const uniqueTemplateSet = new Set(iModelSourceByLocales.values());
+        const iModelFolderPath = path.join(process.cwd(), ResourcesConfig.getInstance().getSkillMetaSrc(profile),
+            CONSTANTS.FILE_PATH.SKILL_PACKAGE.INTERACTION_MODEL, 'custom');
+        // copy iModel JSONs that contains same language
+        iModelSourceByLocales.forEach((v, k) => {
+            if (fs.existsSync(v)) {
+                const sourceLocale = path.basename(v, path.extname(v));
+                const targetFilePath = path.join(iModelFolderPath, `${k}.json`);
+                fs.copySync(v, targetFilePath); // do not fail if the target locale exists already
+                uniqueTemplateSet.delete(v);
+                const sourceManifestLocale = Manifest.getInstance().getPublishingLocale(sourceLocale);
+                Manifest.getInstance().setPublishingLocale(k, sourceManifestLocale);
+            }
+        });
+        // retrieve remote templates when no existing JSON shares the same language
+        _retrieveTemplatesByLanguage(uniqueTemplateSet, doDebug, (templateErr, templateByLanguage) => {
+            if (templateErr) {
+                return callback(templateErr);
+            }
+            iModelSourceByLocales.forEach((v, k) => {
+                if (!fs.existsSync(v)) {
+                    const targetFilePath = path.join(iModelFolderPath, `${k}.json`);
+                    fs.writeFileSync(targetFilePath, templateByLanguage.get(v), { encoding: 'utf-8' });
+                    Manifest.getInstance().setPublishingLocale(k, { name: 'please change' });
+                }
+            });
+            Manifest.getInstance().write();
+            callback(undefined, iModelSourceByLocales);
+        });
+    });
+}
+
+/**
+ * Based on user's selected locales, and the local project's iModel files, decide where to get the
+ * iModel file to add in user's local project. Exclude reusableLocales themselves from the result.
+ *
+ * @param {Array} selectedLocales
+ * @param {String} profile
+ * @param {Boolean} doDebug
+ * @returns {Function} callback Map { locale: filePath/remoteURI }
+ */
+function _getNewLocaleModelUri(selectedLocales, profile, doDebug, callback) {
+    const skillMetaController = new SkillMetadataController({ profile, doDebug });
+    const localIModelByLocale = skillMetaController.getInteractionModelLocales();
+    const result = new Map();
+    // get from map of templates
+    httpClient.request({
+        url: CONSTANTS.TEMPLATES.INTERACTION_MODEL_MAP,
```

Comment:
do you think it is too over complicated having links on s3 that point to gihub urls. There are plus 2 external dependencies. How can an external contributer add/update template?

Reply:
I thought you agreed on this approach already - external won't have permission to add template and it's controlled directly by us. Benefits would be you could easily select template files without linking the selection with certain cli version

Reply:
Sorry. I may misunderstood the solution we agreed on. I just have a few questions/concerns.
1) Do you think it makes sense to put templates also in S3 instead of referencing hello world repo. Why hello world repo is source of truth for this information? If something gets refactored in the hello world repo, I believe it would break our integration.
2) Do you think all those s3 resources should be in some type of repo and have automated process to deploy updates. I think uploading them manually in the UI is error prone (specially it is easy to forget to allow public access). I think this should be taken care before we release  since I don't think it will take much effort to do so.

Reply:
Ok the templates are owned by the templates team, they work on different type of templates catching up with all the new technologies Alexa creates. We currently rely on Github templates for our skill project template as you know.

This iModel JSON file feels to me also rely on them to update, incase there is any new feats introduced into the model json. Also it's consistent for template provisioning. Currently I feel we don't really need a pipeline to poll for the template and cache in our s3 bucket.